### PR TITLE
[Feature] Support pvPortRealloc #1315

### PR DESCRIPTION
**Pull request by skb666**
Description
-----------
This PR introduces a new API `pvPortRealloc` to FreeRTOS, providing dynamic memory reallocation similar to the standard C library's `realloc`. With `configSUPPORT_HEAP_REALLOC` enabled, applications can now resize heap-allocated memory blocks at runtime, either shrinking, expanding in place, or reallocating with data migration when needed.

**Key Changes:**
- Added the `pvPortRealloc` API in `portable/MemMang/heap_4.c` and `heap_5.c`.
- Added the configuration macro `configSUPPORT_HEAP_REALLOC` (default 0, must be 1 to enable realloc support).
- Updated `FreeRTOSConfig.h` and checked for correct configuration use.
- Critical sections and heap integrity checks are enforced throughout realloc logic.
- Follows realloc semantics:
    - NULL pointer behaves like `malloc`.
    - Size 0 behaves like `free`.
    - Tries in-place expand or shrink when possible; otherwise allocates a new block, copies data, and frees the old one.

Test Steps
-----------
1. Enable `configSUPPORT_HEAP_REALLOC` and ensure dynamic allocation is also enabled.
2. Build with either `heap_4.c` or `heap_5.c`.
3. Call `pvPortRealloc` to:
    - Expand, shrink, and reduce memory blocks.
    - Pass `NULL` as input.
    - Pass size `0`.
    - Test fragmentation and edge cases (failing allocation, invalid pointers).
4. Validate heap stats and application logic remain correct.

Checklist:
----------
- [ ] I have tested my changes. No regression in existing tests.
- [ ] I have modified and/or added unit-tests to cover the code changes in this Pull Request.

Related Issue
-----------

#1315 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: include/FreeRTOS.h [hunk 1]
```diff
@@ -2840,6 +2840,14 @@
     #define configSUPPORT_DYNAMIC_ALLOCATION    1
 #endif
 
+#ifndef configSUPPORT_HEAP_REALLOC
+    #define configSUPPORT_HEAP_REALLOC    0
+#endif
+
+#if ( ( configSUPPORT_HEAP_REALLOC > 0 ) && ( configSUPPORT_DYNAMIC_ALLOCATION != 1 ) )
+    #error configSUPPORT_HEAP_REALLOC cannot be used without dynamic allocation, but configSUPPORT_HEAP_REALLOC is not set to 1.
```
Comment (AniruddhaKanhere) [inline include/FreeRTOS.h]:
```suggestion
    #error configSUPPORT_HEAP_REALLOC cannot be used without dynamic allocation, but configSUPPORT_DYNAMIC_ALLOCATION is not set to 1.
```

**Review by AniruddhaKanhere — COMMENTED** [review]
Hello @skb666,

Thank you for taking the time to contribute to the FreeRTOS-kernel.

I have some comments/questions about your PR. Can you please take a look and let me know what you think?

File: portable/MemMang/heap_4.c [hunk 2]
```diff
@@ -410,6 +410,244 @@ void vPortFree( void * pv )
...
+    void *pvReturn = NULL;
+    size_t xOriginalSize;
+    size_t xNewBlockSize;
+    size_t xAdditionalRequiredSize;
+    size_t xCopySize;
+
+    /* Handle NULL pointer case - equivalent to malloc */
+    if( pv == NULL )
+    {
+        return pvPortMalloc( xWantedSize );
```
[the remark this answers is not in the document]
Comment (skb666) [reply portable/MemMang/heap_4.c]: Since the realloc function in the standard library does not follow the single-responsibility principle, the design here aims to maintain a consistent user experience with it. I would like to know whether the refactoring will alter the original behavior of the function.

File: portable/MemMang/heap_4.c [hunk 3]
```diff
@@ -410,6 +410,379 @@ void vPortFree( void * pv )
...
+    size_t xAdditionalRequiredSize;
+    size_t xCurrentBlockSize;
+    size_t xRemainingBlockSize;
+    size_t xNextBlockSize;
+    size_t xPreviousBlockSize;
+    BaseType_t xHasNextBlock;
+    BaseType_t xHasPreviousBlock;
+
+    /* Ensure the end marker has been set up. */
+    configASSERT( pxEnd );
```
Comment (hoxi) [inline portable/MemMang/heap_4.c:464]: This will have to move down to line 472. We can get a pvPortRealloc(NULL, size) as first call; before the heap is initialized.

File: portable/MemMang/heap_5.c [hunk 4]
```diff
@@ -445,6 +445,379 @@ void vPortFree( void * pv )
...
+    size_t xAdditionalRequiredSize;
+    size_t xCurrentBlockSize;
+    size_t xRemainingBlockSize;
+    size_t xNextBlockSize;
+    size_t xPreviousBlockSize;
+    BaseType_t xHasNextBlock;
+    BaseType_t xHasPreviousBlock;
+
+    /* Ensure the end marker has been set up. */
+    configASSERT( pxEnd );
```
Comment (hoxi) [inline portable/MemMang/heap_5.c:499]: This will have to move down to line 507. We can get a pvPortRealloc(NULL, size) as first call; before the heap is initialized.
